iota-trinary/src/trytes.rs

```diff
@@ -0,0 +1,675 @@
+
+impl Add<&Trytes> for Trytes {
+    type Output = Trytes;
+
+    fn add(mut self, rhs: &Trytes) -> Self::Output {
+        self.0.extend_from_slice(rhs.as_bytes());
+        self
+    }
+}
+
+impl AddAssign<&Trytes> for Trytes {
+    fn add_assign(&mut self, rhs: &Trytes) {
+        self.0.extend_from_slice(rhs.as_bytes());
+    }
+}
+
+impl ops::Index<usize> for Trytes {
+    type Output = u8;
+
+    fn index(&self, index: usize) -> &Self::Output {
+        &self[..][index]
+    }
+}
+
+impl ops::Index<ops::Range<usize>> for Trytes {
+    type Output = [u8];
+
+    fn index(&self, index: ops::Range<usize>) -> &Self::Output {
+        &self[..][index]
+    }
+}
+
+impl ops::Index<ops::RangeTo<usize>> for Trytes {
+    type Output = [u8];
+
+    fn index(&self, index: ops::RangeTo<usize>) -> &Self::Output {
+        &self[..][index]
+    }
+}
+
+impl ops::Index<ops::RangeFrom<usize>> for Trytes {
+    type Output = [u8];
+
+    fn index(&self, index: ops::RangeFrom<usize>) -> &Self::Output {
+        &self[..][index]
+    }
+}
+
+impl ops::Index<ops::RangeFull> for Trytes {
+    type Output = [u8];
+
+    fn index(&self, _index: ops::RangeFull) -> &Self::Output {
+        &self.0[..]
+    }
+}
+
+impl ops::Index<ops::RangeInclusive<usize>> for Trytes {
+    type Output = [u8];
+
+    fn index(&self, index: ops::RangeInclusive<usize>) -> &Self::Output {
+        Index::index(&**self, index)
+    }
+}
+
+impl ops::Index<ops::RangeToInclusive<usize>> for Trytes {
+    type Output = [u8];
+
+    fn index(&self, index: ops::RangeToInclusive<usize>) -> &Self::Output {
+        Index::index(&**self, index)
+    }
+}
+
+impl ops::IndexMut<usize> for Trytes {
+    fn index_mut(&mut self, index: usize) -> &mut u8 {
+        &mut self[..][index]
+    }
+}
+
+impl ops::IndexMut<ops::Range<usize>> for Trytes {
+    fn index_mut(&mut self, index: ops::Range<usize>) -> &mut [u8] {
+        &mut self[..][index]
+    }
+}
+
+impl ops::IndexMut<ops::RangeTo<usize>> for Trytes {
+    fn index_mut(&mut self, index: ops::RangeTo<usize>) -> &mut [u8] {
+        &mut self[..][index]
+    }
+}
+
+impl ops::IndexMut<ops::RangeFrom<usize>> for Trytes {
+    fn index_mut(&mut self, index: ops::RangeFrom<usize>) -> &mut [u8] {
+        &mut self[..][index]
+    }
+}
+
+impl ops::IndexMut<ops::RangeFull> for Trytes {
+    fn index_mut(&mut self, _index: ops::RangeFull) -> &mut [u8] {
+        &mut self.0[..]
+    }
+}
+
+impl ops::IndexMut<ops::RangeInclusive<usize>> for Trytes {
+    fn index_mut(&mut self, index: ops::RangeInclusive<usize>) -> &mut [u8] {
+        IndexMut::index_mut(&mut **self, index)
+    }
+}
+
+impl ops::IndexMut<ops::RangeToInclusive<usize>> for Trytes {
+    fn index_mut(&mut self, index: ops::RangeToInclusive<usize>) -> &mut [u8] {
+        IndexMut::index_mut(&mut **self, index)
+    }
+}
+
+impl Deref for Trytes {
+    type Target = [Tryte];
+
+    fn deref(&self) -> &Self::Target {
+        &self.0[..]
+    }
+}
+
+impl DerefMut for Trytes {
```

Comment:
I don't recommend implementing traits like this for `Trytes` at all. `Tryte` is just a type alias of `u8`, which means that it's trivial to modify the structure to contain bytes that do not correspond to valid trytes. If this data structure is to be sound, inputs and outputs should be validated (with the option to disable that validation for release builds).